### PR TITLE
[FG:InPlacePodVerticalScaling] Fix backoff problem when quickly reverting resize patch

### LINKED ISSUE #119838: [FG:InPlacePodVerticalScaling] Container resize policy feature introduced in v1.27 doesn't interrupt CrashLoopBackOff
### What happened?

Hi, does someone know how to “break” the CrashLoopBackOff container status? I don’t want to wait all the backoff time nor delete the whole pod. I’m using the new features of restarting the container after resizing introduced in Kubernetes v1.27, but it seems that after I resize the container using the Python client, it still waits for the back-off time to finish before trying to run the container again.

Deployment info:

```spec:
      containers:
        - name: ...
          image: systemautoscaler/prime-numbers:dev
          imagePullPolicy: Always
          ports:
            - containerPort: ...
          resizePolicy:
            - resourceName: cpu
              restartPolicy: RestartContainer
            - resourceName: memory
              restartPolicy: RestartContainer
```

### What did you expect to happen?

No matter the state of the container, when I resize it, it should be restarted, even if waiting for Back Off.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create a deployment with the spec field as specified above.
2. Scale down container resources to 1m CPU and to 1Mi memory (both limits and requests) using the Kubernetes API e.g. `kubernetes.client.CoreV1Api().patch_namespaced_pod(name=pod_name, namespace=namespace, body=new_pod_data)`
3. Wait till the container is in CrashLoopBackOff and then try to scale up the container resources with enough CPU and memory to run successfully.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.4", GitCommit:"fa3d7990104d7c1f16943a67f11b154b71f6a132", GitTreeState:"clean", BuildDate:"2023-07-19T12:20:54Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"windows/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-15T00:36:28Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}

```

</details>

### Cloud provider

<details>
Working locally on KinD cluster
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# BuildNumber  Caption                                    OSArchitecture  Version
19045        Microsoft Windows 10 Home Single Language  64 bits         10.0.19045
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### LINKED ISSUE #125205: [FG:InPlacePodVerticalScaling] Slow reconcile when quickly reverting resize patch
### What happened?

While working on https://github.com/kubernetes/kubernetes/pull/125202 and testing the second case (patching a pod to perform an in-place resize and then quickly reverting the patch before the resize has been actuated), I've discovered some unexpected behavior. In this case the pod eventually reconciles but it takes about 3 minutes, with the following test output:

```
[sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling] Burstable QoS pod, three containers - no change for c1, increase c2 resources, decrease c3 (net decrease for pod) [sig-node, Feature:InPlacePodVerticalScaling]
k8s.io/kubernetes/test/e2e/node/pod_resize.go:1281
  STEP: Creating a kubernetes client @ 05/29/24 18:13:28.391
  I0529 18:13:28.391963 3291 util.go:499] >>> kubeConfig: /root/kind-test-config
  I0529 18:13:28.393961 3291 util.go:508] >>> kubeContext: kind-kind
  STEP: Building a namespace api object, basename pod-resize @ 05/29/24 18:13:28.394
  STEP: Waiting for a default service account to be provisioned in namespace @ 05/29/24 18:13:28.404
  STEP: Waiting for kube-root-ca.crt to be provisioned in namespace @ 05/29/24 18:13:28.406
  STEP: Creating a kubernetes client @ 05/29/24 18:13:28.408
  I0529 18:13:28.408474 3291 util.go:499] >>> kubeConfig: /root/kind-test-config
  I0529 18:13:28.409838 3291 util.go:508] >>> kubeContext: kind-kind
  STEP: Building a namespace api object, basename pod-resize-resource-quota @ 05/29/24 18:13:28.41
  STEP: Waiting for a default service account to be provisioned in namespace @ 05/29/24 18:13:28.416
  STEP: Waiting for kube-root-ca.crt to be provisioned in namespace @ 05/29/24 18:13:28.417
  STEP: Creating a kubernetes client @ 05/29/24 18:13:28.419
  I0529 18:13:28.419504 3291 util.go:499] >>> kubeConfig: /root/kind-test-config
  I0529 18:13:28.420717 3291 util.go:508] >>> kubeContext: kind-kind
  STEP: Building a namespace api object, basename pod-resize-errors @ 05/29/24 18:13:28.42
  STEP: Waiting for a default service account to be provisioned in namespace @ 05/29/24 18:13:28.427                                                                                                                                                                                                                                                                        
  STEP: Waiting for kube-root-ca.crt to be provisioned in namespace @ 05/29/24 18:13:28.428                                                                                                                                                                                                                                                                                 
  STEP: creating pod @ 05/29/24 18:13:28.429                                                                                                                                                                                                                                                                                                                                
  STEP: verifying the pod is in kubernetes @ 05/29/24 18:13:30.448                                                                                                                                                                                                                                                                                                          
  STEP: verifying initial pod resources, allocations, and policy are as expected @ 05/29/24 18:13:30.453                                                                                                                                                                                                                                                                    
  STEP: verifying initial pod status resources and cgroup config are as expected @ 05/29/24 18:13:30.453                                                                                                                                                                                                                                                                    
  STEP: patching pod for resize @ 05/29/24 18:13:30.457                                                                                                                                                                                                                                                                                                                     
  STEP: verifying pod patched for resize @ 05/29/24 18:13:30.47                                                                                                                                                                                                                                                                                                             
  STEP: patching pod for rollback @ 05/29/24 18:13:30.516                                                                                                                                                                                                                                                                                                                   
  STEP: verifying pod patched for rollback @ 05/29/24 18:13:30.529                                                                                                                                                                                                                                                                                                          
  STEP: waiting for rollback to be actuated @ 05/29/24 18:13:30.529                                                                                                                                                                                                                                                                                                         
  I0529 18:13:30.534186 3291 builder.go:121] Running '/usr/local/bin/kubectl --server=https://127.0.0.1:60962 -- 
 kubeconfig=/root/kind-test-config --context=kind-kind --namespace=pod-resize-4426 exec testpod -- ls 
  /sys/fs/cgroup/cgroup.controllers'
  I0529 18:13:30.679258 3291 builder.go:146] stderr: "Defaulted container \"c1\" out of: c1, c2, c3\n"
  I0529 18:13:30.679299 3291 builder.go:147] stdout: "/sys/fs/cgroup/cgroup.controllers\n"
  I0529 18:13:30.679331 3291 pod_resize.go:379] Namespace pod-resize-4426 Pod testpod Container c1 - looking for cgroup value 209715200 in path /sys/fs/cgroup/memory.max
  I0529 18:13:30.679379 3291 builder.go:121] Running '/usr/local/bin/kubectl --server=https://127.0.0.1:60962 --kubeconfig=/root/kind-test-config --context=kind-kind --namespace=pod-resize-4426 exec testpod --namespace=pod-resize-4426 --container=c1 -- head -n 1 /sys/fs/cgroup/memory.max'
  I0529 18:13:30.819936 3291 builder.go:146] stderr: ""
  I0529 18:13:30.820015 3291 builder.go:147] stdout: "209715200\n"
  I0529 18:13:30.820050 3291 pod_resize.go:379] Namespace pod-resize-4426 Pod testpod Container c1 - looking for cgroup value 20000 100000 in path /sys/fs/cgroup/cpu.max
  I0529 18:13:30.820117 3291 builder.go:121] Running '/usr/local/bin/kubectl --server=https://127.0.0.1:60962 --kubeconfig=/root/kind-test-config --context=kind-kind --namespace=pod-resize-4426 exec testpod --namespace=pod-resize-4426 --container=c1 -- head -n 1 /sys/fs/cgroup/cpu.max'
  I0529 18:13:30.907604 3291 builder.go:146] stderr: ""
  I0529 18:13:30.907650 3291 builder.go:147] stdout: "20000 100000\n"
  I0529 18:13:30.907669 3291 pod_resize.go:379] Namespace pod-resize-4426 Pod testpod Container c1 - looking for cgroup value 4 in path /sys/fs/cgroup/cpu.weight
  I0529 18:13:30.907714 3291 builder.go:121] Running '/usr/local/bin/kubectl --server=https://127.0.0.1:60962 --kubeconfig=/root/kind-test-config --context=kind-kind --namespace=pod-resize-4426 exec testpod --namespace=pod-resize-4426 --container=c1 -- head -n 1 /sys/fs/cgroup/cpu.weight'
  I0529 18:13:30.993819 3291 builder.go:146] stderr: ""
  I0529 18:13:30.993869 3291 builder.go:147] stdout: "4\n"
  I0529 18:13:30.993895 3291 pod_resize.go:379] Namespace pod-resize-4426 Pod testpod Container c2 - looking for cgroup value 314572800 in path /sys/fs/cgroup/memory.max
  I0529 18:13:30.993951 3291 builder.go:121] Running '/usr/local/bin/kubectl --server=https://127.0.0.1:60962 --kubeconfig=/root/kind-test-config --context=kind-kind --namespace=pod-resize-4426 exec testpod --namespace=pod-resize-4426 --container=c2 -- head -n 1 /sys/fs/cgroup/memory.max'
  I0529 18:13:31.078830 3291 builder.go:146] stderr: ""
  I0529 18:13:31.078876 3291 builder.go:147] stdout: "367001600\n"
  I0529 18:13:33.080354 3291 builder.go:121] Running '/usr/local/bin/kubectl --server=https://127.0.0.1:60962 --kubeconfig=/root/kind-test-config --context=kind-kind --namespace=pod-resize-4426 exec testpod --namespace=pod-resize-4426 --container=c2 -- head -n 1 /sys/fs/cgroup/memory.max'
  I0529 18:13:33.205904 3291 builder.go:146] stderr: ""
  I0529 18:13:33.205947 3291 builder.go:147] stdout: "367001600\n"
  I0529 18:13:35.207569 3291 builder.go:121] Running '/usr/local/bin/kubectl --server=https://127.0.0.1:60962 --kubeconfig=/root/kind-test-config --context=kind-kind --namespace=pod-resize-4426 exec testpod --namespace=pod-resize-4426 --container=c2 -- head -n 1 /sys/fs/cgroup/memory.max'
  I0529 18:13:35.335993 3291 builder.go:146] stderr: ""
  I0529 18:13:35.336037 3291 builder.go:147] stdout: "367001600\n"
```

With the last couple of lines continuing indefinitely until it reads the "rolled back" value from `memory.max` in `c2` (3 mins in my case). This also happens inconsistently, you may have to rerun the test case a couple of times to hit it. You can also run the full suite (`-ginkgo.focus="Feature:InPlacePodVerticalScaling"`) to hit it instead of the specific case provided below.


### What did you expect to happen?

After patching forwards and patching backwards, the pod should reach its initial state, and `memory.max` in `c2` should be set to its initial value.

### How can we reproduce it (as minimally and precisely as possible)?

- Check out https://github.com/kubernetes/kubernetes/pull/125202
- Uncomment [this line](https://github.com/kubernetes/kubernetes/pull/125202/files#diff-44a1da1c31a8cd8913e1b8cfb2893c5436bb7c60a4d51c9707f9368bf755ef8fR1391) - `patchAndVerifyAborted`
- Comment both `patchAndVerify` lines (so we only run the "aborted" case), and comment [this line](https://github.com/kubernetes/kubernetes/pull/125202/files#diff-44a1da1c31a8cd8913e1b8cfb2893c5436bb7c60a4d51c9707f9368bf755ef8fR1362)
- Run `e2e.test` with `-ginkgo.focus="Burstable QoS pod, three containers - decrease c1 resources" -num-nodes 2` (you can also run the full suite (`-ginkgo.focus="Feature:InPlacePodVerticalScaling"`) - you may have to run it a couple of times)

### Anything else we need to know?

This could also be an issue with my test changes, so please flag anything that seems suspicious. If you leave `patchAndVerify` uncommented (so patch -> wait for resize -> patch back -> wait for resize) and then run `patchAndVerifyAborted` (on the same case), you will hit a different bug, which I will file separately. 

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.4", GitCommit:"872a965c6c6526caa949f0c6ac028ef7aff3fb78", GitTreeState:"clean", BuildDate:"2022-11-09T13:36:36Z", GoVersion:"go1.19.3", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"31+", GitVersion:"v1.31.0-alpha.0.947+1ff1207d22ab5c-dirty", GitCommit:"1ff1207d22ab5cf442c8dafdf5bded1e32519873", GitTreeState:"dirty", BuildDate:"2024-05-28T19:28:33Z", GoVersion:"go1.22.3", Compiler:"gc", Platform:"linux/arm64"}
WARNING: version difference between client (1.25) and server (1.31) exceeds the supported minor version skew of +/-1
```

</details>


### Cloud provider

<details>
None
</details>


### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
$ uname -a
Linux docker-desktop 5.15.49-linuxkit #1 SMP PREEMPT Tue Sep 13 07:51:32 UTC 2022 aarch64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>


### Install tools

<details>
kind version 0.17.0
</details>


### Container runtime (CRI) and version (if applicable)

<details>

```console
NAME                 STATUS   ROLES           AGE     VERSION                                    INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
kind-control-plane   Ready    control-plane   6h29m   v1.31.0-alpha.0.947+1ff1207d22ab5c-dirty   172.18.0.4    <none>        Ubuntu 22.04.1 LTS   5.15.49-linuxkit   containerd://1.6.9
kind-worker          Ready    <none>          6h29m   v1.31.0-alpha.0.947+1ff1207d22ab5c-dirty   172.18.0.2    <none>        Ubuntu 22.04.1 LTS   5.15.49-linuxkit   containerd://1.6.9
kind-worker2         Ready    <none>          6h29m   v1.31.0-alpha.0.947+1ff1207d22ab5c-dirty   172.18.0.3    <none>        Ubuntu 22.04.1 LTS   5.15.49-linuxkit   containerd://1.6.9
```

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### LINKED ISSUE #125843: [FG:InPlacePodVerticalScaling] Backoff problem when quickly reverting resize patch
### What happened?

While working on #125202 , issue with backoff was discovered resulting to slow reconcile when quickly reverting resize patch. This can be reproduced please check https://github.com/kubernetes/kubernetes/issues/125205#issuecomment-2203184854 ,  https://github.com/kubernetes/kubernetes/issues/125205#issuecomment-2143527472.

Backoff issue does not happen on v1.30 or older please check https://github.com/kubernetes/kubernetes/issues/125205#issuecomment-2203224319

### What did you expect to happen?

After patching forwards and patching backwards a pod which has a container whose `resizePolicy` is `RestartContainer`and a container whose `restartPolicy`is `RestartContainer` , pod should reach its initial state without any backoff delay introduced.

### How can we reproduce it (as minimally and precisely as possible)?

<details>

```console
for l in `seq 301 310`; do kubectl patch pod resize-pod --patch "{\"spec\":{\"containers\":[{\"name\":\"c2\", \"resources\":{\"limits\":{\"memory\":\"${l}Mi\"}}}]}}"; sleep 3; done

$ kubectl get pod resize-pod -w
NAME         READY   STATUS    RESTARTS   AGE
resize-pod   3/3     Running   0          14s
resize-pod   3/3     Running   0          21s
resize-pod   3/3     Running   1 (1s ago)   22s
resize-pod   3/3     Running   1 (3s ago)   24s
resize-pod   3/3     Running   2 (2s ago)   26s
resize-pod   3/3     Running   2 (3s ago)   27s
resize-pod   2/3     CrashLoopBackOff   2 (1s ago)   29s
resize-pod   2/3     CrashLoopBackOff   2 (2s ago)   30s
resize-pod   2/3     CrashLoopBackOff   2 (5s ago)   33s
resize-pod   2/3     CrashLoopBackOff   2 (6s ago)   34s
resize-pod   2/3     CrashLoopBackOff   2 (8s ago)   36s
resize-pod   2/3     CrashLoopBackOff   2 (11s ago)   39s
resize-pod   3/3     Running            3 (13s ago)   41s
```
</details>

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.31.0-alpha.0.19+d591bc54b588b1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.31.0-alpha.2.287+7586ad36136b00
```
</details>

### Cloud provider

<details>
None
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
$ uname -a
# paste output here
Linux sotiris-Precision-5560 6.5.0-41-generic #41~22.04.2-Ubuntu SMP PREEMPT_DYNAMIC Mon Jun  3 11:32:55 UTC 2 x86_64 x86_64 x86_64 GNU/Linux```
</details>

### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by esotsal**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

For container restarts as a result of resize patch, this PR reset backoff for corresponding container.

We need it to overcome slow reconcile ( range of 1-3 minutes) for this corner case. It will still wait for kubelet's next reconcile loop (interval is 1 minute)

#### Which issue(s) this PR fixes:

Fixes #125843

Fixes  #119838

#### Special notes for your reviewer:

Please refer to issue #125205 for background information.

#### Does this PR introduce a user-facing change?

```
NONE
```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -1000,6 +1000,11 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 			reason = reasonStartupProbe
 		} else if isInPlacePodVerticalScalingAllowed(pod) && !m.computePodResizeAction(pod, idx, containerStatus, &changes) {
 			// computePodResizeAction updates 'changes' if resize policy requires restarting this container
+			// overcome slow reconcile ( range of 1-3 minutes ) when reverting resize patch quickly,
+			// before container is started from previous resize patch
+			// it will still wait for kubelet's next reconcile loop (interval is 1 minute), issue 123940
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:1005]: nit: I don't think you need to reference the issue here. If someone wants, they can dig it up from git blame.
```suggestion
			// Resize restart should be effective immediately and clear any accumulated backoff.
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 2]
```diff
@@ -1000,6 +1000,11 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 			reason = reasonStartupProbe
 		} else if isInPlacePodVerticalScalingAllowed(pod) && !m.computePodResizeAction(pod, idx, containerStatus, &changes) {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:1001]: Should the backoff be cleared even if the container doesn't require a restart? I think yes. 2 use cases for this I can think of:
1. Runtime decides to restart anyway (since we don't ever guarantee no restart)
2. Container was stuck in backoff due to oom-kills, so the user updates the container memory limit. It would be nice if the effect was immediate.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 1 above)
Comment (esotsal) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Done

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 2 above)
Comment (esotsal) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:1002]: Thanks for the suggestion, I will need to perform some manual tests and if possible adding tests to capture those corner cases. Will ping you back

Comment (esotsal) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:1002]: @tallclair  I did some tests with your suggestion for the use cases you mention, i think it will require more effort and to create some tests to be more confident. I prefer to create a new issue for items 1 and 2,  keep this for InPlacePodVerticalScaling ( updated title ) to resolve https://github.com/kubernetes/kubernetes/issues/125843 and possible https://github.com/kubernetes/kubernetes/issues/119838. What do you think?

Comment (tallclair) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:1004]: I disagree - why not just address this now? I don't understand your comment about confidence. I don't think it increases the scope much to always restart?

Comment (esotsal) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:1004]: Ok, will give it a try, perhaps i misunderstood what was asked and overestimated the effort.

Comment (esotsal) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: I guess you meant "resize restart" PTAL

**Review by hashim21223445 — APPROVED** [review]

**Comment by esotsal** [comment]
Setting back to WIP, want to create some tests to be more confident

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 2 above)
Comment (esotsal) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: @tallclair i did manual test with scenario 2, works fine with latest patch. Please take another look and let me know. if this is what you meant. Proof below with screencast

https://github.com/user-attachments/assets/8bb38150-c1ac-4574-98ea-3013eb758007

With regard #119838 restart counts increase but container in my lab doesn't return back after being patched with more memory. So i don't think this patch "fully" solves the issue ( will update description )

Comment (esotsal) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Is above "proof" @tallclair @SergeyKanzhelev  sufficient for this commit to be approved or we will need the creation of tests now ?

**Review by tallclair — COMMENTED** [review]
It doesn't feel right to reset the backoff in computePodResizeAction. It also means the backoff doesn't get reset if the container is restarted before getting to computePodReszieAction. How about resetting the backoff in `handlePodResourcesResize` instead? And say we reset it whenever a container's allocated resources changes?

here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2824

**Comment by esotsal** [comment]
> It doesn't feel right to reset the backoff in computePodResizeAction. It also means the backoff doesn't get reset if the container is restarted before getting to computePodReszieAction. How about resetting the backoff in `handlePodResourcesResize` instead? And say we reset it whenever a container's allocated resources changes?
> 
> here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2824

Sounds cleaner, Acknowledged, working on it.

**Comment by esotsal** [comment]
> > It doesn't feel right to reset the backoff in computePodResizeAction. It also means the backoff doesn't get reset if the container is restarted before getting to computePodReszieAction. How about resetting the backoff in `handlePodResourcesResize` instead? And say we reset it whenever a container's allocated resources changes?
> > here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2824
> 
> Sounds cleaner, Acknowledged, working on it.

Done, @tallclair PTAL

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -2822,6 +2822,10 @@ func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod) (*v1.Pod, error) {
 		if err := kl.statusManager.SetPodAllocation(pod); err != nil {
 			return nil, err
 		}
+		for _, container := range pod.Spec.Containers {
+			key := kuberuntime.GetStableKey(pod, &container)
+			kl.backOff.Reset(key)
+		}
```
Comment (tallclair) [inline pkg/kubelet/kubelet.go:2828]: I think we should only reset the backoff if the container is being resized? Can you think of any reason to reset the others?
```suggestion
		for i, container := range pod.Spec.Containers {
			if !apiequality.Semantic.DeepEqual(container.Resources, allocatedPod.Spec.Containers[i].Resources) {
				key := kuberuntime.GetStableKey(pod, &container)
				kl.backOff.Reset(key)
			}
		}
```

Comment (esotsal) [reply pkg/kubelet/kubelet.go]: True, Done. Code is cleaner now, thanks for sharing the DeepEqual method.

**Comment by tallclair** [comment]
Thanks!

BTW, you don't need to specify the tests to rerun. You can just use `/retest` to rerun all the failed tests.

/lgtm
/approve

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 4]
```diff
@@ -1425,7 +1425,7 @@ func (m *kubeGenericRuntimeManager) doBackOff(pod *v1.Pod, container *v1.Contain
 	// Use the finished time of the latest exited container as the start point to calculate whether to do back-off.
 	ts := cStatus.FinishedAt
 	// backOff requires a unique key to identify the container.
-	key := getStableKey(pod, container)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:1428]: is the private version of `getStableKey` needed after this change?

/hold

Comment (SergeyKanzhelev) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:1428]: oh, found it

/unhold

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -2822,6 +2823,12 @@ func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod) (*v1.Pod, error) {
 		if err := kl.statusManager.SetPodAllocation(pod); err != nil {
 			return nil, err
 		}
+		for i, container := range pod.Spec.Containers {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet.go:2826]: another place to support sidecars for https://github.com/kubernetes/kubernetes/pull/128367

@vivzbansal 

I wonder if there is any test that could have catch this.

**Review by SergeyKanzhelev — COMMENTED** [review]
/hold

@tallclair is there any tests we should introduce that demonstrate the bug?

**Comment by tallclair** [comment]
Yeah. @esotsal can you add backoff checking to `TestHandlePodResourcesResize`? https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_test.go#L2580

**Comment by esotsal** [comment]
> Yeah. @esotsal can you add backoff checking to `TestHandlePodResourcesResize`? https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_test.go#L2580

Sure, will give it a try

**Comment by esotsal** [comment]
> > Yeah. @esotsal can you add backoff checking to `TestHandlePodResourcesResize`? https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_test.go#L2580
> 
> Sure, will give it a try

Hi i tried first to test if the fix works manually trying to duplicate the original issue steps as described [here](https://github.com/kubernetes/kubernetes/issues/125205#issuecomment-2143527472). Unfortunately as it is now it is not possible to test because of this issue https://github.com/kubernetes/kubernetes/commit/2bf1f2349cc96cb1cf075489507d905a191119b5#r148798888 . I get Forbidden while attempting the first patch.

**Comment by tallclair** [comment]
Copying comment to here for visibility:

> The patch needs to go to the `/resize` subresource now. If you're trying to test with kubectl, you'll need to build the binary from master, and use the `--subresource=resize` flag.

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (esotsal) [reply pkg/kubelet/kubelet.go:2896]: > another place to support sidecars for #128367

@SergeyKanzhelev inside [SetPodAllocation](https://github.com/Nordix/kubernetes/blob/eea8086d309186daad0644ef94d92a721f16fb60/pkg/kubelet/status/status_manager.go#L247) might another place for #128367

**Comment by esotsal** [comment]
> Copying comment to here for visibility:
> 
> > The patch needs to go to the `/resize` subresource now. If you're trying to test with kubectl, you'll need to build the binary from master, and use the `--subresource=resize` flag.

Thanks i saw the related commit https://github.com/kubernetes/kubernetes/pull/128296 , the example in the help handles one resource but how it should be done for parallel changes as it was possible before?

`kubectl patch deployment nginx-deployment --subresource='scale' --type='merge' -p '{"spec":{"replicas":2}}'`

and the help string

```
   --subresource='':
        If specified, patch will operate on the subresource of the requested object. This flag is beta and may change
        in the future.
```

Can you please share the equivalent for below with the new approach ?  

`./kubectl patch pod resize-pod --subresource='requests' -p '{"spec":{"containers":[{"name":"c1", "resources":{"requests":{"cpu":"50m","memory":"50Mi"},"limits":{"cpu":"150m","memory":"150Mi"}}},{"name":"c2", "resources":{"requests":{"cpu":"350m","memory":"350Mi"},"limits":{"cpu":"450m","memory":"450Mi"}}}]}}'`

**Comment by esotsal** [comment]
> > Copying comment to here for visibility:
> > > The patch needs to go to the `/resize` subresource now. If you're trying to test with kubectl, you'll need to build the binary from master, and use the `--subresource=resize` flag.
> 
> Thanks i saw the related commit #128296 , the example in the help handles one resource but how it should be done for parallel changes as it was possible before?
> 
> `kubectl patch deployment nginx-deployment --subresource='scale' --type='merge' -p '{"spec":{"replicas":2}}'`
> 
> and the help string
> 
> ```
>    --subresource='':
>         If specified, patch will operate on the subresource of the requested object. This flag is beta and may change
>         in the future.
> ```
> 
> Can you please share @tallclair or @AnishShah the equivalent for below with the new approach ?
> 
> ```
> kubectl patch pod resize-pod --patch '{"spec":{"containers":[{"name":"c1", "resources":{"requests":{"cpu":"50m","memory":"50Mi"},"limits":{"cpu":"150m","memory":"150Mi"}}},{"name":"c2", "resources":{"requests":{"cpu":"350m","memory":"350Mi"},"limits":{"cpu":"450m","memory":"450Mi"}}}]}}'   
> sleep 1 
> kubectl patch pod resize-pod --patch '{"spec":{"containers":[{"name":"c1", "resources":{"requests":{"cpu":"100m","memory":"100Mi"},"limits":{"cpu":"200m","memory":"200Mi"}}},{"name":"c2", "resources":{"requests":{"cpu":"200m","memory":"200Mi"},"limits":{"cpu":"300m","memory":"300Mi"}}}]}}'
> ```

Ok found it , equivalent is 

`./kubectl patch pod resize-pod **--subresource='resize'** -p '{"spec":{"containers":[{"name":"c1", "resources":{"requests":{"cpu":"50m","memory":"50Mi"},"limits":{"cpu":"150m","memory":"150Mi"}}},{"name":"c2", "resources":{"requests":{"cpu":"350m","memory":"350Mi"},"limits":{"cpu":"450m","memory":"450Mi"}}}]}}'   `

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -2892,6 +2893,12 @@ func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod, podStatus *kubecontaine
...
```
Comment (liggitt) [inline pkg/kubelet/kubelet.go:2896]: are only main containers resizeable? initContainers / ephemeralContainers aren't?

Comment (tallclair) [reply pkg/kubelet/kubelet.go:2896]: Only main containers & sidecars, but sidecar support hasn't merged yet (coming in https://github.com/kubernetes/kubernetes/pull/128367)

**Comment by tallclair** [comment]
Here's your unit test: https://github.com/tallclair/kubernetes/commit/5533df5385daee246246aa9abc9ec991268b6a3d

Please cherrypick that into your PR so we can unblock this.

**Comment by esotsal** [comment]
> Here's your unit test: [tallclair@5533df5](https://github.com/tallclair/kubernetes/commit/5533df5385daee246246aa9abc9ec991268b6a3d)
> 
> Please cherrypick that into your PR so we can unblock this.

Thanks ! Done
